src/api/procedures/investInSto.ts

```diff
@@ -31,10 +31,53 @@ export type Params = InvestInStoParams & {
  * @hidden
  */
 export interface Storage {
-  investmentPortfolioId: PortfolioId;
+  purchasePortfolioId: PortfolioId;
   fundingPortfolioId: PortfolioId;
 }
 
+/**
+ * @hidden
+ */
+type TierStats = {
+  remaining: BigNumber;
+  price: BigNumber;
+  remainingAmount: BigNumber;
+};
+
+/**
+ * @hidden
+ */
+export const calculateTierStats = (
+  tiers: Tier[],
+  purchaseAmount: BigNumber,
+  maxPrice?: BigNumber
+): TierStats => {
+  return tiers.reduce<TierStats>(
+    (prev, { remaining, price }) => {
+      if ((!maxPrice || price.lte(maxPrice)) && !prev.remainingAmount.isZero()) {
+        if (remaining.gte(prev.remainingAmount)) {
+          return {
+            remaining: prev.remaining.plus(prev.remainingAmount),
+            price: prev.price.plus(prev.remainingAmount.multipliedBy(price)),
+            remainingAmount: new BigNumber(0),
+          };
+        }
+        return {
+          remaining: prev.remaining.plus(remaining),
+          price: prev.price.plus(remaining.multipliedBy(price)),
+          remainingAmount: prev.remainingAmount.minus(remaining),
+        };
```

Comment:
You should be able to avoid using two separate return statements here by assigning the lesser of `remaning` and  `prev.remainingAmount` to a `tierPurchaseAmount` variable and then using it to do all calculations